rust/parquet/src/arrow/array_reader.rs

```diff
@@ -270,91 +268,40 @@ impl<T: DataType> ArrayReader for PrimitiveArrayReader<T> {
+                PhysicalType::INT96
+                | PhysicalType::BYTE_ARRAY
+                | PhysicalType::FIXED_LEN_BYTE_ARRAY => {
+                    // TODO: we could use unreachable!() as this is a private fn
+                    Err(general_err!(
+                        "Cannot read primitive array with a complex physical type"
+                    ))
                 }
-                (ArrowType::Time32(unit), PhysicalType::INT32) => {
-                    match unit {
-                        TimeUnit::Second => {
-                            Time32SecondConverter::new().convert(self.record_reader.cast::<Int32Type>())
-                        }
-                        TimeUnit::Millisecond => {
-                            Time32MillisecondConverter::new().convert(self.record_reader.cast::<Int32Type>())
-                        }
-                        _ => Err(general_err!("Invalid or unsupported arrow array with datatype {:?}", self.get_data_type()))
-                    }
-                }
-                (ArrowType::Time64(unit), PhysicalType::INT64) => {
-                    match unit {
-                        TimeUnit::Microsecond => {
-                            Time64MicrosecondConverter::new().convert(self.record_reader.cast::<Int64Type>())
-                        }
-                        TimeUnit::Nanosecond => {
-                            Time64NanosecondConverter::new().convert(self.record_reader.cast::<Int64Type>())
-                        }
-                        _ => Err(general_err!("Invalid or unsupported arrow array with datatype {:?}", self.get_data_type()))
-                    }
-                }
-                (ArrowType::Interval(IntervalUnit::YearMonth), PhysicalType::INT32) => {
-                    UInt32Converter::new().convert(self.record_reader.cast::<Int32Type>())
-                }
-                (ArrowType::Interval(IntervalUnit::DayTime), PhysicalType::INT64) => {
-                    UInt64Converter::new().convert(self.record_reader.cast::<Int64Type>())
-                }
-                (ArrowType::Duration(_), PhysicalType::INT64) => {
-                    UInt64Converter::new().convert(self.record_reader.cast::<Int64Type>())
-                }
-                (arrow_type, physical_type) => Err(general_err!(
-                    "Reading {:?} type from parquet {:?} is not supported yet.",
-                    arrow_type,
-                    physical_type
-                )),
             }?;
 
+        // cast to Arrow type
+        // TODO: we need to check if it's fine for this to be fallible.
+        // My assumption is that we can't get to an illegal cast as we can only
+        // generate types that are supported, because we'd have gotten them from
+        // the metadata which was written to the Parquet sink
+        let array = arrow::compute::cast(&array, self.get_data_type())?;
```

Comment:
This is the main addition, everything else is to accommodate being able to do this.
The Arrow type was getting erased by the cast converters, so I removed them, and instead preserve `Option<DataType>` along from the schema to get here.
This was also going to become important when dealing with nested lists, as there's a breaking change that we need to make to Arrow in order to correctly roundtrip lists.